### DISCUSSION TITLE
Where are tensor operations defined?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#20820] [Q&A General] [answered]
Hi -

I'm trying to understand more about how MOOSE performs tensor operations but I'm getting lost in all the inheritance stuff going on. Can someone help me find where different methods and operators are defined for tensors? For instance in [RankTwoTensor.h](https://github.com/idaholab/moose/blob/next/framework/include/utils/RankTwoTensor.h), operator overloads are defined in terms of other operator overloads and I can't quite seem to follow the path down to the actual implementation.

Thanks for your help.

**Comment by dschwen** [comment, marked as answer]
`RankTwoTensor` inherits from `RealTensorValue`, which inherits from `TypeTensor`. Both of those are in libmesh. The rank three and four tensors do not inherit from anything and all their operators are implemented in the *Implementation.h files.

**Reply by jessecarterMOOSE** [reply]
Ah, so for instance it looks like RankTwoTensor * RankTwoTensor would bring you [here in type_tensor.h](https://github.com/libMesh/libmesh/blob/devel/include/numerics/type_tensor.h#L1225-L1238)?

**Reply by hugary1995** [reply]
yes
